internal/initializer/waiter.go

```diff
@@ -0,0 +1,83 @@
+/*
+Copyright 2021 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package initializer
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+)
+
+const (
+	errGetCRD             = "cannot get crd"
+	errFmtTimeoutExceeded = "%f seconds timeout for waiting CRDs to be ready is exceeded"
+)
+
+// NewCRDWaiter returns a new *CRDWaiter initializer.
+func NewCRDWaiter(names []string, timeout time.Duration, period time.Duration, log logging.Logger) *CRDWaiter {
+	return &CRDWaiter{Names: names, Timeout: timeout, Period: period, log: log}
+}
+
+// CRDWaiter blocks the execution until all the CRDs whose names are given are
+// deployed to the cluster.
+type CRDWaiter struct {
+	Names   []string
+	Timeout time.Duration
+	Period  time.Duration
+	log     logging.Logger
+}
+
+// Run continuously checks whether the list of CRDs whose names are given are
+// present in the cluster.
+func (cw *CRDWaiter) Run(ctx context.Context, kube client.Client) error {
+	beginning := time.Now()
+	ending := beginning.Add(cw.Timeout)
+	current := beginning
+	cw.log.Info(fmt.Sprintf("Started waiting for the following CRDs to be present: %v", cw.Names))
+	for {
```

Comment:
I think it may be possible to use a `select` here - e.g.

```go
timeout := time.After(cw.Timeout)
ticker := time.NewTicker(cw.Period)
defer ticker.Stop()

for {
	select {
	case <-ticker.C:
		// Main logic goes here. Executed every cw.Period.
	case <-timeout:
		return errors.Errorf(errFmtTimeoutExceeded, cw.Timeout.Seconds())
	}
}
```

Reply:
Nice! I was not aware of that ticker utility. Changed the implementation.